### PR TITLE
feat(tools): add phpstan for PHP static analysis

### LINKED ISSUE #431: feat(tools): add phpstan for PHP static analysis
## Summary

Add PHPStan for PHP static analysis.

## Background

[PHPStan](https://phpstan.org/) is a PHP static analysis tool that finds bugs in code without running it.

## Features

- Finds bugs without running code
- Configurable strictness levels (0-9)
- Type inference without annotations
- Framework-specific extensions (Laravel, Symfony, etc.)
- Custom rule support
- Baseline support for legacy codebases

## Analysis Levels

- Level 0: Basic checks
- Level 5: Type checking
- Level 8: Strictest checks (null safety, etc.)
- Level 9: Maximum strictness

## Why Add

- **PHP is widely used**: Many legacy and modern applications
- **Catches real bugs**: Type errors, null issues, dead code
- **Gradual adoption**: Level system allows incremental strictness
- **Modern PHP**: Encourages type-safe PHP development

## Alternative Consideration

Psalm is another excellent PHP static analyzer. Could consider adding both or choosing one.

## Tasks

- [ ] Create tool definition in `lintro/tools/definitions/phpstan.py`
- [ ] Create parser in `lintro/parsers/phpstan/`
- [ ] Add tests
- [ ] Add documentation

## References

- https://phpstan.org/
- https://github.com/phpstan/phpstan

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title: `feat(tools): add phpstan for PHP static analysis`

- Type:
  - [x] feat (minor)

- Breaking change:
  - [ ] none

## What's Changing

Adds [PHPStan](https://phpstan.org/) as a check-only lintro tool plugin for PHP
static analysis, following the existing plugin pattern (ToolDefinition,
BaseToolPlugin, native parser, manifest/version registration, install surfaces,
docs, and tests).

## Parser choice: native JSON (not shared SARIF)

Per the fidelity checklist in `docs/design/sarif-ingestion-evaluation.md`, this uses
a **native** parser on `--error-format=json`:

- **PHPStan does not emit SARIF.** Its formatters are `raw, gitlab, table, junit,
  checkstyle, teamcity, github, json, prettyJson` — no first-class SARIF, so the
  shared SARIF path is not applicable without an external converter.
- **SARIF would drop PHPStan-specific fidelity even if available.** Each message
  carries an `identifier` (e.g. `function.notFound`, drives the doc URL
  `https://phpstan.org/error-identifiers/<identifier>`) and an optional `tip`
  (remediation hint). SARIF has no clean slot for these (identifier depends on the
  tool populating `rules[].helpUri`, which PHPStan does not; `tip` would be lost in
  a generic `properties` bag). The native parser preserves `identifier`, `tip`,
  `ignorable`, file, line, and message.

## Config / level decision: run-with-defaults

PHPStan requires an analysis `level` and has no built-in default. Decision (mirrors
ruff/rubocop running with defaults, not stylelint/vale skipping):

- **No native config** (`phpstan.neon` / `phpstan.neon.dist` / `phpstan.dist.neon`):
  inject `--level=0`, the most conservative level. It flags only unambiguous bugs
  (undefined symbols, wrong argument counts) that do not depend on an autoloader or
  annotations, keeping false positives low on standalone files.
- **Native config present**: it defines the level, so lintro does not pass `--level`
  and defers entirely to the project config.

Verified a bare `.php` file (no autoloader) analyzes safely without crashing.

## Details

- `stdout`/`stderr` are separated (`_run_subprocess_result`) so PHPStan's stderr
  AI-guidance preamble never corrupts JSON parsing; pass/fail is decided by the
  parsed issue count (PHPStan exits non-zero on findings).
- Shared registration (minimal/additive): `tool_name`, `doc_url_template`,
  `_tool_versions` (2.2.5), `manifest.json` (+ `php` language map), `version_parsing`
  / `version_checking` install hint, a `renovate.json` custom manager, and the
  `pyproject.toml` parser package.
- Install surfaces: `install-tools.sh` (phpstan.phar), `Dockerfile` (php-cli + both
  verify blocks), Homebrew template (`depends_on "phpstan"` + caveats). Homebrew
  phpstan is 2.2.5 — matches the pin.
- Docs: README, getting-started, configuration, and a tool-analysis writeup.

## Testing strategy

- Parser unit tests on real captured PHPStan output + edge cases (empty/null/invalid
  JSON, missing keys, malformed messages, top-level errors).
- Plugin unit tests (mocked subprocess): definition, doc_url, option validation,
  level injection vs. native-config suppression, check violations/clean, fix raises.
- Binary-gated integration tests on PHP fixtures with seeded errors (skip when the
  PHP toolchain is absent).

Full suite: `6030 passed, 10 skipped` (2 known env-only pre-existing failures
deselected). `ruff`, `ruff format`, `mypy` (phpstan files), `shellcheck`, `shfmt`,
`hadolint`, and `markdownlint-cli2` all clean on changed files. `lintro doctor`
reports `phpstan 2.2.5 (>= 2.2.5)`. Manifest generator `--check` reports no drift.

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [x] Docs updated
- [x] Local checks passed (full pytest suite + native linters on changed files)

## Related Issues

Closes #431

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds PHPStan support for PHP static analysis. The main changes are:

- A new PHPStan plugin and native JSON parser.
- Tool registration, version metadata, manifest entries, and install support.
- PHP documentation, site navigation, sample fixtures, and tests.
</details>

<h3>Confidence Score: 4/5</h3>

This is close, but the installer verification should be fixed before merging.
- The PHPStan plugin and parser paths look consistent.
- Explicit PHPStan installs now fail when PHP is missing.
- The default installer can still report PHPStan as available when only a stale PHAR exists without PHP.

**Files Needing Attention:** scripts/utils/install-tools.sh

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lintro/tools/definitions/phpstan.py | Adds PHPStan command construction, config handling, subprocess execution, and parsed result handling. |
| lintro/parsers/phpstan/phpstan_parser.py | Adds parsing for PHPStan JSON file messages and top-level errors. |
| scripts/utils/install-tools.sh | Adds PHPStan installation and verification logic, with a remaining runtime check gap in verification. |

</details>

<a href="https://app.greptile.com/api/ide/cursor?prompt=Greploop%20lgtm-hq%2Fpy-lintro%20PR%20%231160%3A%20work%20through%20Greptile's%20open%20review%20comments%2C%20then%20keep%20reviewing%20and%20fixing%20until%20it%20comes%20back%20clean%20at%205%2F5%20with%20zero%20unresolved%20comments.%0AStart%20by%20reading%20the%20comments%20off%20the%20PR%20itself.%20On%20GitHub%2C%20use%20paginated%20%60gh%20api%20graphql%60%20to%20query%20%60pullRequest.reviewThreads%60%20with%20each%20thread's%20%60isResolved%60%20value%20and%20inline%20%60comments%60.%20%60gh%20pr%20view%20--comments%60%20only%20includes%20conversation%20comments%2C%20so%20do%20not%20use%20it%20as%20the%20findings%20source.%20They%20are%20not%20listed%20in%20this%20prompt%20on%20purpose%3A%20the%20PR%20is%20current%2C%20a%20pasted%20copy%20would%20not%20be.%20Skip%20anything%20already%20resolved%2C%20and%20if%20you%20judge%20a%20comment%20wrong%2C%20say%20so%20rather%20than%20changing%20code%20to%20satisfy%20it.%0A%0APrefer%20the%20Greptile%20CLI%2C%20which%20reviews%20the%20working%20tree%20with%20no%20push%20and%20no%20CI%20run.%20Fall%20back%20to%20PUSH%20LOOP%20only%20where%20a%20step%20below%20says%20to.%0A1.%20Run%20%60command%20-v%20greptile%60.%20Missing%3A%20go%20to%20PUSH%20LOOP%2C%20and%20pitch%20the%20CLI%20at%20the%20end.%0A2.%20Run%20%60greptile%20whoami%60.%20It%20exits%200%20even%20when%20signed%20out%2C%20so%20read%20the%20output%3A%20if%20it%20says%20%22Not%20signed%20in%22%2C%20ask%20the%20user%20to%20run%20%60greptile%20login%60%20%28it%20opens%20a%20browser%2C%20do%20not%20run%20it%20yourself%29%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%0A3.%20Look%20for%20%60greploop%2FSKILL.md%60%20under%20%60.agents%2Fskills%2F%60%20or%20%60.claude%2Fskills%2F%60%2C%20both%20in%20the%20repo%20root%20and%20in%20the%20user's%20home%20directory.%20Found%3A%20follow%20that%20skill%2C%20never%20pushing%20between%20rounds%2C%20and%20you%20are%20done.%0A4.%20Missing%3A%20run%20%60greptile%20skills%20list%60.%20If%20it%20fails%20as%20an%20unknown%20command%2C%20this%20CLI%20predates%20the%20skill%2C%20so%20tell%20the%20user%20to%20run%20%60greptile%20update%60%20and%20go%20to%20PUSH%20LOOP%20with%20no%20pitch.%20Otherwise%20ask%20the%20user%20before%20running%20%60greptile%20skills%20install%20greploop%60%2C%20then%20follow%20the%20skill%20it%20writes.%0A%0APUSH%20LOOP%20%28only%20when%20a%20step%20above%20sends%20you%20here%29%3A%0A%0AGREPLOOP%20ENABLED%3A%20After%20fixing%20all%20issues%2C%20commit%20and%20push%20your%20changes.%20Greptile%20will%20automatically%20re-review%20the%20PR%20via%20webhook.%20Wait%20for%20the%20review%20to%20complete%2C%20then%20read%20the%20new%20review%20comments%20and%20confidence%20score.%20If%20the%20confidence%20score%20is%20below%205%2F5%20or%20there%20are%20unresolved%20comments%2C%20fix%20those%20issues%20and%20push%20again.%20Repeat%20until%20the%20confidence%20score%20is%205%2F5%20with%20zero%20unresolved%20comments.%20Maximum%205%20iterations.%20Do%20NOT%20open%20a%20new%20PR%20%E2%80%94%20keep%20pushing%20to%20the%20same%20branch.%0A%0AIF%20YOU%20ENDED%20UP%20ON%20PUSH%20LOOP%20BECAUSE%20THE%20CLI%20WAS%20MISSING%3A%20once%20the%20loop%20is%20done%2C%20print%20the%20block%20below%20and%20nothing%20else%20about%20the%20CLI.%20Reproduce%20it%20character%20for%20character%20inside%20a%20fenced%20code%20block%2C%20so%20the%20borders%20line%20up.%20Do%20not%20reword%20it%2C%20translate%20it%2C%20summarize%20it%2C%20add%20to%20it%2C%20or%20drop%20the%20box.%20Print%20it%20once%2C%20and%20do%20not%20bring%20the%20CLI%20up%20again%20later%20in%20the%20session.%0A%0A%E2%95%AD%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%95%AE%0A%E2%94%82%20%20You%20can%20run%20greploops%20faster%20locally%20with%20our%20CLI.%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%E2%94%82%0A%E2%94%82%20%20Install%20it%20at%20https%3A%2F%2Fwww.greptile.com%2Fcli%2C%20or%20I%20can%20install%20it%20for%20you.%20%20%E2%94%82%0A%E2%95%B0%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%94%80%E2%95%AF%0A%0AIf%20they%20take%20you%20up%20on%20it%2C%20install%20with%20%60npm%20install%20-g%20greptile%60%20%28or%20%60brew%20install%20greptileai%2Ftap%2Fgreptile%60%29%2C%20then%20%60greptile%20skills%20install%20greploop%60.%20Leave%20%60greptile%20login%60%20to%20them%2C%20it%20opens%20a%20browser.&pr=1160&platform=github">Fix all with Greploop</a> <a href="https://app.greptile.com/api/ide/cursor?prompt=%23%23%23%20Issue%201%0Ascripts%2Futils%2Finstall-tools.sh%3A1887%0A**Verify%20PHP%20runtime**%0A%0AWhen%20the%20default%20install%20runs%20on%20a%20machine%20with%20an%20old%20%60phpstan%60%20PHAR%20on%20%60PATH%60%20but%20no%20%60php%60%20runtime%2C%20the%20install%20block%20skips%20PHPStan%20because%20PHP%20is%20missing.%20This%20verification%20list%20still%20checks%20only%20that%20%60phpstan%60%20exists%2C%20so%20the%20script%20can%20report%20success%20even%20though%20the%20PHAR%20cannot%20run.%20Later%20%60lintro%20check%20--tools%20phpstan%60%20fails%20when%20it%20tries%20to%20execute%20the%20tool%20without%20PHP.%20The%20PHPStan%20verification%20should%20require%20the%20runtime%20too%2C%20or%20skip%20PHPStan%20verification%20when%20the%20install%20path%20skipped%20it.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&pr=1160&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=6">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=%23%23%23%20Issue%201%0Ascripts%2Futils%2Finstall-tools.sh%3A1887%0A**Verify%20PHP%20runtime**%0A%0AWhen%20the%20default%20install%20runs%20on%20a%20machine%20with%20an%20old%20%60phpstan%60%20PHAR%20on%20%60PATH%60%20but%20no%20%60php%60%20runtime%2C%20the%20install%20block%20skips%20PHPStan%20because%20PHP%20is%20missing.%20This%20verification%20list%20still%20checks%20only%20that%20%60phpstan%60%20exists%2C%20so%20the%20script%20can%20report%20success%20even%20though%20the%20PHAR%20cannot%20run.%20Later%20%60lintro%20check%20--tools%20phpstan%60%20fails%20when%20it%20tries%20to%20execute%20the%20tool%20without%20PHP.%20The%20PHPStan%20verification%20should%20require%20the%20runtime%20too%2C%20or%20skip%20PHPStan%20verification%20when%20the%20install%20path%20skipped%20it.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=lgtm-hq%2Fpy-lintro&pr=1160&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22lgtm-hq%2Fpy-lintro%22%20on%20the%20existing%20branch%20%22feat%2F431-add-phpstan%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22feat%2F431-add-phpstan%22.%0A%0A%23%23%23%20Issue%201%0Ascripts%2Futils%2Finstall-tools.sh%3A1887%0A**Verify%20PHP%20runtime**%0A%0AWhen%20the%20default%20install%20runs%20on%20a%20machine%20with%20an%20old%20%60phpstan%60%20PHAR%20on%20%60PATH%60%20but%20no%20%60php%60%20runtime%2C%20the%20install%20block%20skips%20PHPStan%20because%20PHP%20is%20missing.%20This%20verification%20list%20still%20checks%20only%20that%20%60phpstan%60%20exists%2C%20so%20the%20script%20can%20report%20success%20even%20though%20the%20PHAR%20cannot%20run.%20Later%20%60lintro%20check%20--tools%20phpstan%60%20fails%20when%20it%20tries%20to%20execute%20the%20tool%20without%20PHP.%20The%20PHPStan%20verification%20should%20require%20the%20runtime%20too%2C%20or%20skip%20PHPStan%20verification%20when%20the%20install%20path%20skipped%20it.%0A%0A---%0A%0AFor%20each%20issue%20above%2C%20determine%20whether%20it%20is%20valid%20and%20should%20be%20fixed.%20If%20so%2C%20fix%20it%20directly.&repo=lgtm-hq%2Fpy-lintro&pr=1160&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=6">Fix All in Codex</picture></a>

<sub>Reviews (9): Last reviewed commit: ["fix(phpstan): align with main completene..."](https://github.com/lgtm-hq/py-lintro/commit/923e7ad4bfe6a92d5fe52d0c689eb109206db10c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=42247680)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: lintro/tools/definitions/phpstan.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+        PHPStan auto-discovers ``phpstan.neon`` / ``phpstan.neon.dist`` /
+        ``phpstan.dist.neon`` from the current working directory. When one is
+        present it defines the analysis level, so lintro must not also pass
+        ``--level`` (which would either be redundant or conflict).
+
+        Returns:
+            True when a native config file exists in the current directory.
+        """
+        cwd = Path.cwd()
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/definitions/phpstan.py]: Fixed in the pushed commit: _has_native_config now takes the subprocess's run cwd (ctx.cwd) and checks for phpstan.neon/.dist there, so --level suppression and PHPStan's own config discovery are evaluated against the same directory.

File: lintro/tools/definitions/phpstan.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                issues_count=0,
+            )
+
+        return ToolResult(
+            name=self.definition.name,
+            success=issues_count == 0,
+            output=None,
+            issues_count=issues_count,
+            issues=issues,
+        )
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/definitions/phpstan.py:287]: Fixed: the guard no longer requires empty stdout — any non-zero exit with zero parsed issues fails the run, so partial JSON or a PHP fatal error on stdout cannot pass as clean. Test added.

File: scripts/utils/install-tools.sh [hunk 3]
```diff
@@ -750,6 +750,29 @@ main() {
...
+		echo -e "${BLUE}Installing phpstan...${NC}"
+		PHPSTAN_VERSION=$(get_tool_version "phpstan") || exit 1
+		if [ $DRY_RUN -eq 1 ]; then
+			log_info "[DRY-RUN] Would install phpstan v${PHPSTAN_VERSION}"
+		elif command -v phpstan &>/dev/null; then
+			echo -e "${GREEN}✓ phpstan already installed${NC}"
+		else
+			phar_url="https://github.com/phpstan/phpstan/releases/download/${PHPSTAN_VERSION}/phpstan.phar"
+			if download_with_retries "$phar_url" "$BIN_DIR/phpstan" 3; then
+				chmod +x "$BIN_DIR/phpstan"
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/utils/install-tools.sh:1016]: Fixed: the installer now checks for a php interpreter before downloading and skips with a clear install hint when absent, instead of leaving an unrunnable PHAR on PATH for verification to trip over.

File: scripts/utils/install-tools.sh [hunk 4]
```diff
@@ -750,6 +750,35 @@ main() {
...
+		if [ $DRY_RUN -eq 1 ]; then
+			log_info "[DRY-RUN] Would install phpstan v${PHPSTAN_VERSION}"
+		elif command -v phpstan &>/dev/null; then
+			echo -e "${GREEN}✓ phpstan already installed${NC}"
+		elif ! command -v php &>/dev/null; then
+			# The PHAR cannot run (or be verified) without a PHP interpreter;
+			# installing it anyway would fail the later verification step and
+			# leave a broken binary on PATH.
+			echo -e "${YELLOW}⚠ Skipping phpstan: no 'php' interpreter on PATH." \
+				"Install PHP (apt install php-cli / brew install php) and re-run.${NC}"
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/utils/install-tools.sh]: Fixed in the pushed commit: the missing-php branch now exits 1 with the install hint instead of warning and continuing — an explicit phpstan request without its runtime is an error.

File: scripts/utils/install-tools.sh [hunk 5]
```diff
@@ -751,6 +751,36 @@ main() {
...
+		elif command -v phpstan &>/dev/null && command -v php &>/dev/null; then
+			echo -e "${GREEN}✓ phpstan already installed${NC}"
+		elif ! command -v php &>/dev/null; then
+			# The PHAR cannot run (or be verified) without a PHP interpreter;
+			# installing it anyway would leave a broken binary on PATH and a
+			# failing verification. An explicit request for phpstan without
+			# its runtime is an error, not a skip.
+			echo -e "${RED}✗ Cannot install phpstan: no 'php' interpreter on PATH." \
+				"Install PHP (apt install php-cli / brew install php) and re-run.${NC}"
+			exit 1
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/utils/install-tools.sh]: Fixed: the installer no longer aborts the whole run when PHP is missing on the default all-tools path. When `--tools` explicitly requests phpstan (`TOOL_FILTER` set) and no `php` interpreter is on PATH it still errors; otherwise it prints a yellow skip hint ("Install PHP to enable it") and continues installing the remaining tools.

**Comment by TurboCoder13** [comment]
Parked (2026-09-04, owner decision) — converted to draft and labelled `blocked`.

The tool-definition layer is being factored (#2311 in the re-alignment roadmap #2288): shared check/fix scaffolding is lifted into `lintro/tools/core`, definitions move to one package per tool, and a duplicate-code gate (#2293) now fails any PR that adds another copy of the current template. This PR was written against the current template and would add ~300 duplicated lines plus ~30 central edits.

When #2311's acceptance criteria hold, a comment will land here with the new shape; the executing agent then rebases this PR if the conflict set is smaller than the diff, or closes it and re-implements from its issue. No action needed until then.

**Comment by TurboCoder13** [comment]
Phase 2 of #2288 (un-park after C.3 #2311): the package-per-tool layout is now on `main` (e0f3786c). A tool lives in `lintro/tools/<tool>/{__init__.py, definition.py}` (no `lintro/tools/definitions/` anymore); discovery scans per-tool packages; `lintro/plugins/_builtin_index.py` is regenerated with `python3 scripts/ci/generate-builtin-tool-index.py`; the package must be appended to `[tool.lintro.pylint] include` and `GATE_PACKAGES` in tests/unit/test_duplicate_code_baseline.py (duplicate-code gate is zero-tolerance); every definition declares `claims`/`reads_tree`/`partitionable` (#1740, PR #2420, landing next); shared per-file/batch runners in `lintro/tools/core/{check_runner,fix_runner,batch_runner}.py` replace hand-rolled loops. See docs/contributing/adding-a-new-tool.md. This PR will be rebased onto that layout one at a time, or closed and re-implemented from its issue if the rebase is larger than the original diff.
